TypeCobol/Compiler/Symbols/TypedVariableSymbol.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System.IO;
+using TypeCobol.Compiler.Types;
+
+namespace TypeCobol.Compiler.Symbols
+{
+    /// <summary>
+    /// A variable declared with a type that comes from a TypeDef.
+    /// Such variable is expanded to have the expanded type from its TypeDef.
+    /// </summary>
+    public class TypedVariableSymbol : VariableSymbol
+    {
+        /// <summary>
+        /// Constructor with an unresolved Type's path
+        /// </summary>
+        /// <param name="name">Variable's name</param>
+        /// <param name="paths">The unresolved type's path</param>
+        public TypedVariableSymbol(string name, string[] paths)
+            : base(name)
+        {
+            System.Diagnostics.Debug.Assert(paths != null);
+            System.Diagnostics.Debug.Assert(paths.Length != 0);
+            SetFlag(Flags.HasATypedefType, true);
+            TypePaths = paths;
+            Typedef = null;
+        }
+
+        /// <summary>
+        /// If the underlying type is not resolved then this is the path of the type to resolve.
+        /// </summary>
+        public string[] TypePaths { get; }
+
+        /// <summary>
+        /// The Typedef symbol
+        /// </summary>
+        /// <remarks>Setter should only be called by TypeCobolLinker.</remarks>
+        public TypedefSymbol Typedef { get; set; }
+
+        /// <summary>
+        /// The Type of a variable whose type comes from a typedef is only accessible after the typedef has been resolved.
+        /// It is then replaced by a new Type instance during program expansion.
+        /// </summary>
+        public override Type Type
+        {
+            get
+            {
+                if (Typedef == null)
+                {
+                    //Type linking not done yet
+                    return null;
+                }
+
+                if (base.Type == null)
+                {
+                    //Type expansion not done yet 
+                    return Typedef.Type;
```

Comment:
All information is there but... In your example what we get here is a TypedVariableSymbol with a TypedefType targeting Date and a base.Type of ArrayType (with no ElementType set).
All is recombined during expansion and in the end the variable will have an ArrayType with fields from Date.
But I agree this could be improved by prioritizing 'array' over 'type' in such declaration, we should get TypedefType targeting ArrayType, itself targeting the date type. Expansion would need to be modified too to accomodate with this new model.

Reply:
My previous comment is wrong, `TypedefType` is for `TypedefSymbol` not for `TypedVariableSymbol`.
So we have all the information both in `TypePaths` on `TypedVariableSymbol` and in its `base.Type` which would be an `ArrayType`.

Another way to model this is to introduce a new Type, `TypeReference` or `UnresolvedType` which would hold the TypedefPath. The TypeCobolLinker would replace every UnresolvedType with its corresponding TypedefType coming from the resolved typedef. With this model, we could also get rid of `TypedVariableSymbol` class. (a typed variable is merely a VariableSymbol with initially an UnresolvedType).
What are your thoughts @osmedile and @mayanje ?

Reply:
We already have SymbolReference to represent a reference to something (like a type).
TypeCobolLinker can already do its jobs without UnresolvedType. We don't need to introduce any new concept.

As this PR is only focused on IBM Enterprise Cobol 6, we can discussed about Type in another PR.

Reply:
I agree with this last point of view, thinking that a variable with an UnresolvedType targets a variable having a  type form a TypeDef is a simplification, because UnresolvedType means that in presence of a type inference algorithm we didn't succeed to infer a type. So I agree to let the current implementation as it is, don't introduce this new concept.

Reply:
Ok